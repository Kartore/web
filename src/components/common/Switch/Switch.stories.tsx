import type { Meta, StoryObj } from '@storybook/react';

import { Switch } from '.';

const meta = {
	component: Switch,
} satisfies Meta<typeof Switch>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
	args: {},
};
