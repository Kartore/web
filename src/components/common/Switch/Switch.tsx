import type { FC } from 'react';
import { useRef } from 'react';

import type { AriaSwitchProps } from 'react-aria';
import { VisuallyHidden, useSwitch } from 'react-aria';
import { useToggleState } from 'react-stately';

import { cn } from '~/utils/tailwindUtil';

export type SwitchProps = {
  className?: string;
  label: string;
} & AriaSwitchProps;

export const Switch: FC<SwitchProps> = ({ className, label, ...props }) => {
  const state = useToggleState(props);
  const ref = useRef(null);
  const { labelProps, inputProps } = useSwitch(props, state, ref);
  return (
    <div className={cn('flex flex-row items-center justify-between', className)}>
      <label
        {...labelProps}
        className={cn('text-sm font-semibold text-gray-600', labelProps.className)}
      >
        {label}
      </label>
      <label className={'cursor-pointer'}>
        <p className={'sr-only'}>{label}</p>
        <VisuallyHidden>
          <input {...inputProps} ref={ref} />
        </VisuallyHidden>
        <svg width={40} height={24} aria-hidden="true" style={{ marginRight: 4 }}>
          <rect
            x={4}
            y={4}
            width={32}
            height={16}
            rx={8}
            fill={state.isSelected ? 'orange' : 'gray'}
          />
          <circle cx={state.isSelected ? 28 : 12} cy={12} r={5} fill="white" />
        </svg>
      </label>
    </div>
  );
};

Switch.displayName = 'Switch';
