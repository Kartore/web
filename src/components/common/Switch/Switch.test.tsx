import { composeStories } from '@storybook/react';
import { render } from '@testing-library/react';
import { describe } from 'vitest';

import * as Stories from './Switch.stories';

const { Default } = composeStories(Stories);

describe('Component: Switch', () => {
  describe('Snapshot', () => {
    it('Default', () => {
      const { asFragment } = render(<Default />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
