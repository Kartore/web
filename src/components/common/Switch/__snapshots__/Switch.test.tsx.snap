// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component: Switch > Snapshot > Default 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row items-center justify-between"
  >
    <label
      class="text-sm font-semibold text-gray-600"
    />
    <label
      class="cursor-pointer"
    >
      <p
        class="sr-only"
      />
      <div
        style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
      >
        <input
          role="switch"
          type="checkbox"
        />
      </div>
      <svg
        aria-hidden="true"
        height="24"
        style="margin-right: 4px;"
        width="40"
      >
        <rect
          fill="gray"
          height="16"
          rx="8"
          width="32"
          x="4"
          y="4"
        />
        <circle
          cx="12"
          cy="12"
          fill="white"
          r="5"
        />
      </svg>
    </label>
  </div>
</DocumentFragment>
`;
